src/category_theory/subterminal.lean

```diff
@@ -0,0 +1,131 @@
+/-
+Copyright (c) 2020 Bhavik Mehta. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Bhavik Mehta
+-/
+import category_theory.limits.shapes.terminal
+import category_theory.limits.shapes.binary_products
+
+/-!
+# Subterminal objects
+
+Subterminal objects are the objects which can be thought of as subobjects of the terminal object.
+In fact, the definition can be constructed to not require a terminal object, by defining `A` to be
+subterminal iff for any `Z`, there is at most one morphism `Z ⟶ A`.
+An alternate definition is that the diagonal morphism `A ⟶ A ⨯ A` is an isomorphism.
+In this file we define subterminal objects and show the equivalence of these three definitions.
+
+We also construct the subcategory of subterminal objects.
+
+## TODO
+
+* Define exponential ideals, and show this subcategory is an exponential ideal.
+* Define subobject lattices in general, show that `subterminals C` is equivalent to the subobject
+category of a terminal object.
+* Use both the above to show that in a locally cartesian closed category, every subobject lattice
+  is cartesian closed (equivalently, a Heyting algebra).
+
+-/
+universes v₁ v₂ u₁ u₂
+
+noncomputable theory
+
+namespace category_theory
+
+open limits category
+
+variables {C : Type u₁} [category.{v₁} C] {A : C}
+
+/-- An object `A` is subterminal iff for any `Z`, there is at most one morphism `Z ⟶ A`. -/
+def is_subterminal (A : C) : Prop := ∀ ⦃Z : C⦄ (f g : Z ⟶ A), f = g
```

Comment:
Is there a reason that you don't use `subsingleton` in this def?

Reply:
I thought that subsingleton was intended to be used as a class rather than a prop - I don't mind switching though

Reply:
Hmm, let's leave it like this. I don't think it matters much in practice.

Reply:
Oops - last night I had a go at changing, but I forgot to push the updated version. I agree it doesn't make a huge difference, the main changes were:

```
/-- An object `A` is subterminal iff for any `Z`, there is at most one morphism `Z ⟶ A`. -/
def is_subterminal (A : C) : Prop := ∀ (Z : C), subsingleton (Z ⟶ A)

lemma is_subterminal.hom_ext (hA : is_subterminal A) {Z : C} (f g : Z ⟶ A) : f = g :=
by { haveI := hA Z, apply subsingleton.elim }
```
and then to prove something's subterminal, the proof looks like 
```
/--
If the unique morphism from `A` to a terminal object is a monomorphism, `A` is subterminal.
The converse of `is_subterminal.mono_is_terminal_from`.
-/
lemma is_subterminal_of_mono_is_terminal_from {T : C} (hT : is_terminal T) [mono (hT.from A)] :
  is_subterminal A :=
λ Z, ⟨λ f g, by { rw ← cancel_mono (hT.from A), apply hT.hom_ext }⟩
```

I suppose that since the exposed API is basically the same either way, it shouldn't make much difference like you say.